### ISSUE TITLE
incomplete or wrong error message when executing ambiguous Ex command

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#9564]
**Steps to reproduce**

Run this shell command:

    vim -Nu NONE -S <(tee <<'EOF'
        vim9script
        command CmdA echomsg 'CmdA'
        command CmdB echomsg 'CmdB'
        nnoremap <F3> <ScriptCmd>Cmd<CR>
        feedkeys("\<F3>")
    EOF
    )

`E464` is given with this message:

    E464: Ambiguous use of user-defined command

**Expected behavior**

`E464` is given with this message:

    E464: Ambiguous use of user-defined command:     Cmd
                                                     ^^^

Notice the presence of the name of the ambiguous Ex command.

**Version of Vim**

8.2 Included patches: 1-4103

**Environment**

Operating system: Ubuntu 20.04.3 LTS
Terminal: xterm
Value of $TERM: xterm-256color
Shell: zsh 5.8

**Additional context**

No issue if the ambiguous command is executed from a script, instead of being executed interactively:
```vim
vim9script
command CmdA echomsg 'CmdA'
command CmdB echomsg 'CmdB'
Cmd
```
    E464: Ambiguous use of user-defined command: Cmd
                                                 ^^^

However, in compiled code, we get the same issue again:
```vim
vim9script
command CmdA echomsg 'CmdA'
command CmdB echomsg 'CmdB'
def Func()
    Cmd
enddef
defcompile
```
    E464: Ambiguous use of user-defined command

---

At the script level, the error message might be wrong or misleading:
```vim
vim9script
command CmdA echomsg 'CmdA'
command CmdB echomsg 'CmdB'
Cmd#xxx
```
    E464: Ambiguous use of user-defined command: Cmd#xxx

The message gives the wrong impression that `#xxx` is not ignored.  It is.  If it was not, there would be no ambiguity.  Besides, the name of a command cannot include `#`:
```vim
vim9script
command Cmd#xxx eval 0 + 0
```
    E182: Invalid command name

BTW, it would help if `E182` could print the name of the invalid command, just like `E464` can print the name of the ambiguous command.

**Comment by lacygoill** [comment]
> I think this example is not realistic.

It is realistic, because I found it while debugging another issue, and I found the message confusing; i.e. with a better message, I might have found the fix faster.

I have this mapping:

    xnoremap <unique> <Tab> <C-\><C-N><ScriptCmd>:* call UltiSnips#SaveLastVisualSelection()<CR>gvs

Since it's defined in a Vim9 script, I thought I could get rid of `call` (thanks to `<ScriptCmd>`), but it doesn't work: it gives the previously mentioned error.

I think that because of the `:*` range, what follows is not parsed as a function call, but a command name.  And a command name cannot include `#`.  So, the latter is parsed as the start of an inline comment, which makes everything afterward ignored.  As a result, Vim tries to execute this:

    :vim9 :* UltiSnips

But I have several commands starting with `UltiSnips`.

    :com UltiSnips
        Name              Args Address Complete    Definition
        UltiSnipsAddFiletypes 1                    :call UltiSnips#AddFiletypes(<q-args>)
    !   UltiSnipsEdit     ?            customlist  :call UltiSnips#Edit(<q-bang>, <q-args>)

Hence the `E464` error.

But the real issue had nothing to do with an ambiguous command; it was the `:*` range which changed the semantics of the next token.  But I need that range, because the function is a legacy one defined with the `range` attribute.  IOW, I really need `:call`.

   > This is a corner case.

OK, fair enough.

**Comment by lacygoill** [comment]
> In other situations it can be useful to show more than what was recognized as the command.

This would be a better message:

    E464: Ambiguous use of user-defined command: Cmd in "Cmd#xxx"
                                                     ^----------^
                                                       context

And it still gives us the context in which the ambiguous command was found.

**Comment by lacygoill** [comment]
> Well, it does include the line number.

When run at the script level, yes, but not when run from an autocmd.

**Comment by brammool** [comment]
>    > I think this example is not realistic.
> 
> It is realistic, because I found it while debugging another issue, and
> I found the message confusing; i.e. with a better message, I might
> have found the fix faster.
> 
> I have this mapping:
> 
>     xnoremap <unique> <Tab> <C-\><C-N><ScriptCmd>:* call UltiSnips#SaveLastVisualSelection()<CR>gvs
> 
> Since it's defined in a Vim9 script, I thought I could get rid of
> `call` (thanks to `<ScriptCmd>`), but it doesn't work: it gives the
> previously mentioned error.

I was referring to using "Cmd#xxx", which suggests a user defined
command "Cmd" followed by "#xxx".  You are actually using an autoload
script name.


> I think that because of the `:*` range, what follows is not parsed as
> a function call, but a command name.  And a command name cannot
> include `#`.  So, the latter is parsed as the start of an inline
> comment, which makes everything afterward ignored.  As a result, Vim
> tries to execute this:
> 
>     :vim9 :* UltiSnips

Yeah, figuring out what is a command and what is something else
(function name, variable) has become quite complicated.  Leaving out
"call" and "let" has made the syntax nicer, but figuring out what the
line contains much more complicated.

But here the problem seems to be that after skipping a range the rules
for detecting a command are different.  This is intentionally done in
the code, and I have trouble remembering why...

If I remove the condition, then this line fails:

	:substitute(some(other(

Thus ":" not only indicates a range, but also implies using legacy
script syntax.  This is actually documented: 

	In the rare case there is ambiguity between a function name and an Ex
	command, prepend ":" to make clear you want to use the Ex command.

But it might be unexpected if you only wanted to use a range and
otherwise keep everything the same.  I'm afraid I don't have an idea of
how to make this work better.  I'll add a note in the documentation.

> But I have several commands starting with `UltiSnips`.
> 
>     :com UltiSnips
>         Name              Args Address Complete    Definition
>         UltiSnipsAddFiletypes 1                    :call UltiSnips#AddFiletypes(<q-args>)
>     !   UltiSnipsEdit     ?            customlist  :call UltiSnips#Edit(<q-bang>, <q-args>)
> 
> Hence the `E464` error.
> 
> But the real issue had nothing to do with an ambiguous command; it was
> the `:*` range which changed the semantics of the next token.  But I
> need that range, because the function is a legacy one defined with the
> `range` attribute.  IOW, I really need `:call`.

I'm afraid I can't think of something to give a better error message.
At this point Vim just doesn't know what was intended.

-- 
Mental Floss prevents moral decay!

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by lacygoill** [comment]
There are still 2 things we could do to improve these error messages.

First, if Vim finds out an ambiguous command, *and* it's followed by some characters, then, in the error message, it could separate the name of the command from the rest.  We had the exact same issue in #8149.  If we had a global `:Foo` command and a buffer-local `:Foo3bar` command installed at some later point, executing `:Foo3bar` gave this error message:

    E1144: Command is not followed by white space: Foo3bar

This was confusing.  The issue was fixed in [8.2.2811](https://github.com/vim/vim/releases/tag/v8.2.2811).  Now, Vim gives this error message:

    E1144: Command "Foo" is not followed by white space: Foo3bar

This is much better.  Same thing here.  This wrong command:

    :* UltiSnips#SaveLastVisualSelection()

Gives this error message:

    E464: Ambiguous use of user-defined command: :* UltiSnips#SaveLastVisualSelection()

This is not easy to understand.  I asked Vim to call a function, not to execute a command.  So why does it complain about a command?   This is a rhetorical question, I know the answer.  But this confusion could be avoided with a message such as:

    E464: Ambiguous "UltiSnips" command: UltiSnips#SaveLastVisualSelection()

Now it's easier to understand.  Vim did not parse `UltiSnips#SaveLastVisualSelection()` as a function, but as a command followed by an inline comment.

---

Second, why do we even allow a command to be followed by an inline comment?
```vim
vim9script
var dir = '/tmp/.vim'
&runtimepath = dir
dir ..= '/autoload'
dir->mkdir('p')
var lines =<< trim END
    vim9script
    def Script#func()
    enddef
END
lines->writefile(dir .. '/Script.vim')
command ScriptA eval 1 + 1
:% Script#func()
```
    E1144: Command "Script" is not followed by white space: Script#func()

This error message is easy to understand, and Vim did not accept `:Script` to be followed by an inline comment.
Now, watch what happens if I define a second Ex command, `:ScriptB`, so that `:Script` becomes ambiguous:

    E464: Ambiguous use of user-defined command: :% Script#func()
```vim
vim9script
var dir = '/tmp/.vim'
&runtimepath = dir
dir ..= '/autoload'
dir->mkdir('p')
var lines =<< trim END
    vim9script
    def Script#func()
    enddef
END
lines->writefile(dir .. '/Script.vim')
command ScriptA eval 1 + 1
command ScriptB eval 2 + 2
:% Script#func()
```
    E464: Ambiguous use of user-defined command: :% Script#func()

This message is harder to understand, but more importantly, why does Vim now allow an inline comment to follow `:Script`?  It did not accept that in the previous snippet.  Either it should be always allowed, or never.

According to [`:help vim9-differences`](https://vimhelp.org/vim9.txt.html#vim9-differences), it should be never:

   > To improve readability there must be a space between a command and the #
   > that starts a comment: >
   >         var name = value # comment
   >         var name = value# error!
